### PR TITLE
Warmup byepoch

### PR DESCRIPTION
**Pull request by kennymckormick**
Feature: warmup byepoch
Summary: Sometimes warmup should last n epochs. Since we may use different batchsize for training, the exact number of iterations may change. To use warmup byepoch, when building lr_updater, you need to set `warmup_byepoch=True`.

### CODE REVIEW

**Comment by hellock** [comment]
There are some unexpected files included.

File: mmcv/runner/hooks/lr_updater.py [hunk 1]
```diff
@@ -12,6 +12,7 @@ def __init__(self,
                  warmup=None,
                  warmup_iters=0,
                  warmup_ratio=0.1,
+                 warmup_byepoch=False,
```
Comment (hellock) [inline mmcv/runner/hooks/lr_updater.py]: `warmup_by_epoch`

File: mmcv/runner/runner.py [hunk 2]
```diff
@@ -336,6 +337,15 @@ def run(self, data_loaders, workflow, max_epochs, **kwargs):
...
                          get_host_info(), work_dir)
         self.logger.info('workflow: %s, max: %d epochs', workflow, max_epochs)
+
+        # set epoch length, in case we use warmup_byepoch
+        # only one tuple in can be with mode 'train'
+        for i, flow in enumerate(workflow):
+            mode, epochs = flow
+            if mode == 'train':
+                self.epoch_len = len(data_loaders[i])
+                break
```
Comment (hellock) [inline mmcv/runner/runner.py]: No need to set `epoch_len` here, we can set it in the hook to reduce the influence to the runner.

File: mmcv/runner/hooks/lr_updater.py [hunk 3]
```diff
@@ -62,6 +64,9 @@ def before_run(self, runner):
         self.base_lr = [
             group['initial_lr'] for group in runner.optimizer.param_groups
         ]
+        if self.warmup_byepoch:
+            assert runner.epoch_len
+            self.warmup_iters *= runner.epoch_len
```
Comment (hellock) [inline mmcv/runner/hooks/lr_updater.py]: This can be implemented in `before_train_epoch`, where we can obtain the epoch_len by `len(runner.data_loader)`.

File: mmcv/runner/runner.py [hunk 4]
```diff
@@ -254,6 +255,7 @@ def train(self, data_loader, **kwargs):
         self.model.train()
         self.mode = 'train'
         self.data_loader = data_loader
+        self.active_loader = data_loader
```
Comment (hellock) [inline mmcv/runner/runner.py]: What is the difference from `self.data_loader` and `self.active_loader`? It may not be necessary to assign a new attribute here.

File: mmcv/runner/hooks/lr_updater.py [hunk 5]
```diff
@@ -29,6 +30,12 @@ def __init__(self,
         self.warmup = warmup
         self.warmup_iters = warmup_iters
         self.warmup_ratio = warmup_ratio
+        self.warmup_by_epoch = warmup_by_epoch
+
+        if self.warmup_by_epoch:
+            self.warmup_epochs = self.warmup_iters
+            self.warmup_iters = None
```
Comment (hellock) [inline mmcv/runner/hooks/lr_updater.py]: Is it necessary?

File: mmcv/runner/runner.py (hunk 4 above)
Comment (kennymckormick) [reply mmcv/runner/runner.py]: Sorry, didn't see that. My bad.
